### PR TITLE
Feature/location categories

### PR DESCRIPTION
**Pull request by rma-bryson**
Adds Location Groups and Categories with test classes.

### CODE REVIEW

**Review by rma-psmorris — CHANGES_REQUESTED** [review]
rebase to main please

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationController.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package mil.army.usace.hec.cwms.radar.client.controllers;
```
Comment (rma-psmorris) [inline cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationController.java:1]: this file isnt related to your change

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationEndpointInput.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package mil.army.usace.hec.cwms.radar.client.controllers;
```
Comment (rma-psmorris) [inline cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationEndpointInput.java:1]: this file isnt related to your change

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationGroupEndpointInput.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * SOFTWARE.
+ */
+
+package mil.army.usace.hec.cwms.radar.client.controllers;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import org.junit.jupiter.api.Test;
+
+class TestLocationGroupEndpointInput {
```
Comment (rma-psmorris) [inline cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationGroupEndpointInput.java:31]: test operation methods

File: cwms-radar-client/src/test/resources/radar/json/location.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (rma-psmorris) [inline cwms-radar-client/src/test/resources/radar/json/location.json:1]: not related to this change

File: cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/AssignedLocation.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
+package mil.army.usace.hec.cwms.radar.client.model;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.swagger.annotations.ApiModelProperty;
+
+@javax.annotation.Generated(value = "io.swagger.codegen.v3.generators.java.SpringCodegen", date = "2021-11-02T12:25:34.578-07:00[America/Los_Angeles]")
+public class AssignedLocation {
```
Comment (rma-psmorris) [inline cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/AssignedLocation.java:7]: Are these all of the attributes supported by the DB table?

File: cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/Location.java [hunk 6]
```diff
@@ -0,0 +1,366 @@
+package mil.army.usace.hec.cwms.radar.client.model;
```
Comment (rma-psmorris) [inline cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/Location.java:1]: not related to this change

File: cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/LocationGroup.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
+package mil.army.usace.hec.cwms.radar.client.model;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.swagger.annotations.ApiModelProperty;
+import java.util.List;
+
+public class LocationGroup {
+
+    @JsonProperty("id")
```
Comment (rma-psmorris) [inline cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/LocationGroup.java:9]: are these all of the db columns?

File: cwms-radar-client/src/test/resources/radar/json/catalog_ts.json [hunk 8]
```diff
@@ -0,0 +1,3007 @@
+{
```
Comment (rma-psmorris) [inline cwms-radar-client/src/test/resources/radar/json/catalog_ts.json:1]: not related to current change

File: cwms-radar-client/src/test/resources/radar/json/catalog_tspage1.json [hunk 9]
```diff
@@ -0,0 +1,3007 @@
+{
```
Comment (rma-psmorris) [inline cwms-radar-client/src/test/resources/radar/json/catalog_tspage1.json:1]: not related to current change

File: cwms-radar-client/src/test/resources/radar/json/catalog_tspage2.json [hunk 10]
```diff
@@ -0,0 +1,3007 @@
+{
```
Comment (rma-psmorris) [inline cwms-radar-client/src/test/resources/radar/json/catalog_tspage2.json:1]: not related to current change

**Review by rma-psmorris — CHANGES_REQUESTED** [review]
Review on loc categories api

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationGroupEndpointInput.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    public LocationGroupEndpointInput categoryId(String categoryId) {
+        this.categoryId = categoryId;
+        return this;
+    }
+
+    @Override
+    protected HttpRequestBuilder addInputParameters(HttpRequestBuilder httpRequestBuilder) {
+        return httpRequestBuilder.addQueryParameter("group-id", groupId)
```
[the remark this answers is not in the document]
Comment (rma-bryson) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationGroupEndpointInput.java:53]: I went off the order I saw on swagger for location groups. If ordering it office, category, group makes more sense here, I'll update it.

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationCategoryController.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+     * Retrieve Location Category.
+     *
+     * @param apiConnectionInfo             - connection info
+     * @param locationCategoryEndpointInput - category id and office
+     * @return LocationCategory
+     * @throws IOException - thrown if retrieve failed
+     */
+    public LocationCategory retrieveLocationCategory(ApiConnectionInfo apiConnectionInfo, LocationCategoryEndpointInput locationCategoryEndpointInput)
+        throws IOException {
+        HttpRequestResponse response = new HttpRequestBuilder(apiConnectionInfo, "location/category")
```
[the remark this answers is not in the document]
Comment (rma-bryson) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationCategoryController.java:46]: updated to use constant

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationGroupController.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * @param apiConnectionInfo          - connection info
+     * @param locationGroupEndpointInput - group id, office id, and category id for location group
+     * @return LocationGroup based on inputs given
+     * @throws IOException - thrown if retrieval failed
+     */
+    public LocationGroup retrieveLocationGroup(ApiConnectionInfo apiConnectionInfo, LocationGroupEndpointInput locationGroupEndpointInput)
+        throws IOException {
+        HttpRequestResponse response = new HttpRequestBuilder(apiConnectionInfo, "location/group")
```
[the remark this answers is not in the document]
Comment (rma-bryson) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationGroupController.java:44]: updated to use constant

File: cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/LocationGroup.java (hunk 7 above)
Comment (rma-bryson) [reply cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/LocationGroup.java:9]: I more want off of what RADAR was using, but on double-checking, that looks to be all the db columns

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationCategoryEndpointInput.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+ * SOFTWARE.
+ */
+
+package mil.army.usace.hec.cwms.radar.client.controllers;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import org.junit.jupiter.api.Test;
+
+class TestLocationCategoryEndpointInput {
```
[the remark this answers is not in the document]
Comment (rma-bryson) [reply cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationCategoryEndpointInput.java:35]: Added operations test

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationGroupEndpointInput.java (hunk 3 above)
Comment (rma-bryson) [reply cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationGroupEndpointInput.java:36]: added test for operation

File: cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/AssignedLocation.java (hunk 5 above)
Comment (rma-bryson) [reply cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/AssignedLocation.java:7]: Yes

File: cwms-radar-client/src/test/resources/radar/json/catalog_tspage2.json (hunk 10 above)
Comment (rma-bryson) [reply cwms-radar-client/src/test/resources/radar/json/catalog_tspage2.json]: These were removed on rebase.

File: cwms-radar-client/src/test/resources/radar/json/catalog_tspage1.json (hunk 9 above)
Comment (rma-bryson) [reply cwms-radar-client/src/test/resources/radar/json/catalog_tspage1.json]: These were removed on rebase.

File: cwms-radar-client/src/test/resources/radar/json/catalog_ts.json (hunk 8 above)
Comment (rma-bryson) [reply cwms-radar-client/src/test/resources/radar/json/catalog_ts.json]: These were removed on rebase.

File: cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/Location.java (hunk 6 above)
Comment (rma-bryson) [reply cwms-radar-model/src/main/java/mil/army/usace/hec/cwms/radar/client/model/Location.java:1]: Once PR6 is merged, i can do another rebase off main, which will remove this

File: cwms-radar-client/src/test/resources/radar/json/location.json (hunk 4 above)
Comment (rma-bryson) [reply cwms-radar-client/src/test/resources/radar/json/location.json]: Once PR6 is merged, i can do another rebase off main, which will remove this

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationEndpointInput.java (hunk 2 above)
Comment (rma-bryson) [reply cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationEndpointInput.java:25]: Once PR6 is merged, i can do another rebase off main, which will remove this

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationController.java (hunk 1 above)
Comment (rma-bryson) [reply cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationController.java:25]: Once PR6 is merged, i can do another rebase off main, which will remove this

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationEndPointInput.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
+package mil.army.usace.hec.cwms.radar.client.controllers;
```
[the remark this answers is not in the document]
Comment (rma-bryson) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationEndPointInput.java:25]: Once PR6 is merged, i can do another rebase off main, which will remove this

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationController.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
+package mil.army.usace.hec.cwms.radar.client.controllers;
```
[the remark this answers is not in the document]
Comment (rma-bryson) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationController.java:25]: Once PR6 is merged, i can do another rebase off main, which will remove this

**Review by rma-psmorris — APPROVED** [review]
approve

**Pull request merged**